File: SecondYearCS/OOP/Civitas/src/civitas/MazoSorpresa.java
package civitas;
import java.util.ArrayList;
import java.util.Collections;
        
public class MazoSorpresa {
    
    private boolean barajada;
    private boolean debug;
    private Sorpresa ultimaSorpresa;
    private int usadas;

    private ArrayList<Sorpresa> sorpresas;
    private ArrayList<Sorpresa> cartasEspeciales; 
    // La variable cartasEspeciales almacenará la carta sorpresa del tipo SALIRCARCEL 
    // mientras se considere retirada del mazo (y por tanto en posesión de un jugador). 
    // Se ha definido como una colección para poderle dar más funcionalidad al juego
    
    //Inicializar varias variables privadas
    //para no repetir código en los diferentes constructores
    final void init(){
        this.barajada = false;
        this.usadas = 0;
        this.sorpresas = new ArrayList<>();
        this.cartasEspeciales = new ArrayList<>();
    }
    
    //Constructor sin parámetros
    MazoSorpresa(){
        this.debug = false;
        this.init();
    }
    
    //Constructor con parámetros para this.debug = d
    MazoSorpresa(boolean d){
        this.debug = d;
        this.init();
        if( debug ){
            Diario.getInstance().ocurreEvento("Debug");
        }
    }
    
    /**
     * @brief Si el mazo no ha sido barajado añadimos la nueva sorpresa.
     * @param s nueva sorpresa a añadir
     */
    void alMazo(Sorpresa s){
        if(!this.barajada){
            this.sorpresas.add(s);
        }
    }
    
    /**
     * @brief En caso de haber utilizado todas cartas, o no haberlas barajadas,
     * debemos entonces barajar las cartas utilizando @b Collections.shuffle();
     * @return ultimaSorpresa.
     */
    Sorpresa siguiente(){
        if((!this.barajada || this.usadas==this.sorpresas.size() )&& !this.debug){
            Collections.shuffle(sorpresas);
            this.usadas = 0;
            this.barajada = true;
        }
        this.usadas++;
        this.ultimaSorpresa = this.sorpresas.get(0);//Al quitar una carta
        this.sorpresas.add(this.ultimaSorpresa); //La añadimos al final de la pila
        this.sorpresas.remove(0);
        
        return ultimaSorpresa;       
    }    
    
    /**
     * @brief Si @b this.sorpresas posee el parámetro, entonces lo retiramos y
     * lo añadimos a las @b cartasEspeciales
     * @param sorpresa carta a quitar del mazo
     */
    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        if (this.sorpresas.contains(sorpresa)){
            this.cartasEspeciales.add(sorpresa);
            this.sorpresas.remove(this.sorpresas.indexOf(sorpresa));
            Diario.getInstance().ocurreEvento("InhabilitarEspecial");
        }
    }
    
    /**
     * @brief Exactamente la inversa de inhabilitarCartaEspecial()
     * @param sorpresa 
     */
    void habilitarCartaEspecial(Sorpresa sorpresa){
        if (this.cartasEspeciales.contains(sorpresa)){
            this.sorpresas.add(sorpresa);
            this.cartasEspeciales.remove(this.cartasEspeciales.indexOf(sorpresa));
            Diario.getInstance().ocurreEvento("HabilitarEspecial");
        }
    }
}

File: SecondYearCS/OOP/CivitasRB/lib/mazo_sorpresa.rb
module CivitasRB
  require_relative 'sorpresa.rb'
  require_relative 'tipo_sorpresa.rb'
  class MazoSorpresa
    #Setter/getters de las variables privadas necesarias.
    attr_accessor :barajada, :usadas, :debug, :ultimaSorpresa 
    #Constructor con parámetro con valor por defecto
    def initialize(debug=false)
      @ultimaSorpresa = 0
      @barajada = false
      @usadas = 0
      @debug = debug
      if @debug 
        Diario.instance.ocurre_evento("Debug Activated in Mazo")
      end
      @sorpresas = Array.new
      @cartasEspeciales = Array.new
    end
    
    #Si no hemos barajado el mazo, podemos seguir añadiendo cartas.
    def alMazo(s)
      if !@barajado
        @sorpresas << s
      end
    end
    
    #Si no hemos barajado el mazo, o ya hemos utilizado todas las cartas
    #Entonces debemos barajar el mazo y luego sacar una de las cartas
    #Si ya estabamos jugando, entonces simplemente sacamos la siguiente
    #carta
    def siguiente
      if ( !@barajado || @sopresas.length == @usadas ) && !@debug
        @sorpresas = @sorpresas.shuffle #Contrario al .sort
        @usadas = 0
        @barajada = true
      end
      @usadas += 1
      @ultimaSorpresa = @sorpresas.at(0)
      @sorpresas.delete_at(0)
      @sorpresas << @ultimaSorpresa
      return @ultimaSorpresa
    end
    
    #Si tenemos en la baraja una carta "s", entonces debemos quitarla
    #Y añadirla en la baraja de cartasEspeciales.
    def inhabilitarCartaEspecial(s)
      if @sorpresas.include?(s)
        @sorpresas.delete(s)
        @cartasEspeciales << s
        Diario.instance.ocurre_evento("Hemos desactivado una carta especial")
      end
    end
    
    #Verificamos si cartasEspeciales posee una carta "s", en caso
    #afirmativo debemos quitarle esa carta y añadirle a nuestra baraja normal.
    def habilitarCartaEspecial(s)
      if @cartasEspeciales.include?(s)
        @cartasEspeciales.delete(s)
        @sorpresas << s
        Diario.instance.ocurre_evento("Hemos activado una carta especial")
      end
    end
    
  end
end
